ci-operator/step-registry/cucushift/upgrade/check/capability/cucushift-upgrade-check-capability-commands.sh

```diff
@@ -205,32 +205,80 @@ function extract_migrated_caps() {
     done
 }
 
-# Get the capabilities which are introdcued for the 1st time in the target payload
-# By comparing two payloads at manifest level
-function extract_potential_new_caps() {
-    local source_dir="$1"
-    local target_dir="$2"
-    local target_cap_files target_file filename source_file target_cap
-    local candidate_new_caps=""
+function lowerVersion() {
+    local version="$1" target_version="$2"
 
-    # Check for missing environment variables
-    if [ -z "$source_dir" ] || [ -z "$target_dir" ]; then
-        echo "ERROR: Missing required arguments"
-        return 1
+    local ret=0
+    if [[ "$(printf '%s\n' "${target_version}" "${version}" | sort --version-sort | head -n1)" == "${target_version}" ]]; then
+        ret=1
     fi
+    return $ret
+}
 
-    target_cap_files=$(grep -rl "capability.openshift.io/name" ${target_dir})
-    for target_file in ${target_cap_files}; do
-        filename=$(basename "${target_file}")
-        source_file="$source_dir/$filename"
+# Get the new caps transfered from an existing operator on source verion
+# by removing those cap operators which is totally unknow on source version
+function get_transfered_new_caps() {
+    local candidate_new_caps="$1" source_ver="$2"
+    local cap tmp_candidate_new_caps
 
-        if [ ! -f "$source_file" ]; then
-            target_cap=$(grep 'capability.openshift.io/name:' "$target_file" | cut -d ':' -f2 | tr -d '"' | sort | uniq )
-            candidate_new_caps="${candidate_new_caps} ${target_cap}"
+    tmp_candidate_new_caps="${candidate_new_caps}"
+    source_ver=$(echo "$source_ver" | cut -d. -f1,2)
+    for cap in ${candidate_new_caps}; do
+        #shellcheck disable=SC2076
+        if [[ " ${!new_caps_version[*]} " =~ " ${cap} " ]] && lowerVersion "${source_ver}" "${new_caps_version[$cap]}"; then
```

Comment:
when will `${source_ver} > ${new_caps_version[$cap]}`?

Reply:
Good question. In the current real world, there is no such case.

But I can foresee a case in the future. 
1. Assumed on 4.N version, a new operator is introduced and GAed. Let us call it "A". On the 4.N+1 version, the operator become an optional operator. `new_caps_version["A"]="4.N"`
2. Assumed there is no the condition `lowerVersion "${source_ver}" "${new_caps_version[$cap]}"` in this check.
3. In 4.N -> 4.N+1 upgrade case, `candidate_new_caps` is "A", after the function is executed, `transfered_new_caps` is set to null. It will lead to a failed result (A is installed in the upgraded cluster, but the testing is expecting it is disabled).
4. In 4.N-1 -> 4.N+1 upgrade case, `candidate_new_caps` is "A", after the function is executed, `transfered_new_caps` is set to null. It will lead to a passed result.

In the above point 2, if we have the `lowerVersion` condition.
3.  In 4.N -> 4.N+1 upgrade case, `candidate_new_caps` is "A", after the function is executed, `transfered_new_caps` is set to "A". It will lead to a passed result.
4.  In 4.N-1 -> 4.N+1 upgrade case, `candidate_new_caps` is "A", after the function is executed, `transfered_new_caps` is set to null. It will lead to a passed result.

Reply:
> Assumed on 4.N version, a new operator is introduced and GAed. Let us call it "A". On the 4.N+1 version, the operator become an optional operator.

In this scenario, "A" does not meet first introduced operational cap, right? So "A" should not be defined in new_caps_version[], but only added into caps_string of get_caps_for_version_capset(). 
Then, in 4.N -> 4.N+1 upgrade case, candidate_new_caps is "A", after the function is executed, transfered_new_caps is still set to "A". It will lead to a passed result.

In 4.N-1 -> 4.N+1 upgrade case, candidate_new_caps is "A", after the function is executed, transfered_new_caps is still set to "A". It will lead to a passed result. "A" will be enabled.

Reply:
> In 4.N-1 -> 4.N+1 upgrade case, candidate_new_caps is "A", after the function is executed, transfered_new_caps is still set to "A". It will lead to a passed result. "A" will be enabled.

For 4.N-1, the resources belonging to "A" would not be installed, after the upgrade, there would be no any resources belonging to "A" installed, even the cluster is on 4.N+1 version, then the testing would be failed, because the script is expecting "A" was enabled.

Reply:
4.N-1 -> 4.N+1 will go to 4.N first, right? "A" is a new operator but not an optional cap in 4.N, so during 4.N-1-> 4.N, "A" will be installed after upgrade to 4.N by default. And then during 4.N->4.N+1, "A" is still installed.

Reply:
Even go to 4.N first, "A" would not be installed automatically, though it is not an optional cap in 4.N.

Reply:
Hmm, if "A" is a new operator(non optional) in 4.N but not 4.N-1, will "A" be installed or not if setting `BaselineCapabilitySet=None` during installation? I just checked the api definition in [doc](https://github.com/openshift/api/blob/6f735e7109c87826edee3ca02a753071ecc933b9/config/v1/types_cluster_version.go#L297), `BaselineCapabilitySet=None` seems only control optional caps. So I guess "A" is not in the optional caps, then it should be installed, right?

Reply:
Maybe you are right, in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-stable-4.12-upgrade-from-stable-4.11-aws-c2s-ipi-disc-priv-fips-f60/1863851873665028096/artifacts/aws-c2s-ipi-disc-priv-fips-f60/, `control-plane-machine-set` is introduced in 4.12 for the 1st time, after upgrade to 4.12 from 4.11, the co is installed in the upgraded cluster.

Let me update the script.

Reply:
Done, I keep `lowerVersion` function in case of possible demand in the future.